python/pyarrow/_dataset.pyx

```diff
@@ -42,6 +43,52 @@ def _forbid_instantiation(klass, subclasses_instead=True):
     raise TypeError(msg)
 
 
+ctypedef CResult[shared_ptr[CRandomAccessFile]] CCustomOpen()
+
+cdef class FileSource:
+
+    cdef:
+        # XXX why is shared_ptr necessary here? CFileSource shouldn't need it
+        shared_ptr[CFileSource] wrapped
```

Comment:
@pitrou @kszucs I tried using 
```suggestion
        CFileSource wrapped
```
which caused a segfault after `FileSource.__init__()`. Am I missing something obvious?

Reply:
using __cinit__ instead of __init__ seems to fix it locally